RecoTracker/MkFit/plugins/MkFitInputConverter.cc

```diff
@@ -0,0 +1,211 @@
+                                      const mkfit::LayerNumberConverter& lnc) const {
+  for (const auto& detset : hits) {
+    const DetId detid = detset.detId();
+    const auto subdet = detid.subdetId();
+    const auto layer = ttopo.layer(detid);
+    const auto isStereo = ttopo.isStereo(detid);
+
+    for (const auto& hit : detset) {
+      if (!passCCC(hit, detid))
+        continue;
+
+      const auto& gpos = hit.globalPosition();
+      SVector3 pos(gpos.x(), gpos.y(), gpos.z());
+      const auto& gerr = hit.globalPositionError();
+      SMatrixSym33 err;
+      err.At(0, 0) = gerr.cxx();
+      err.At(1, 1) = gerr.cyy();
+      err.At(2, 2) = gerr.czz();
+      err.At(0, 1) = gerr.cyx();
+      err.At(0, 2) = gerr.czx();
+      err.At(1, 2) = gerr.czy();
+
+      const auto ilay = lnc.convertLayerNumber(subdet, layer, false, isStereo, gpos.z() > 0);
+      LogTrace("MkFitInputConverter") << "Adding hit detid " << detid.rawId() << " subdet " << subdet << " layer "
+                                      << layer << " isStereo " << isStereo << " zplus " << (gpos.z() > 0) << " ilay "
+                                      << ilay;
+
+      indexLayers.insert(hit.firstClusterRef().id(), hit.firstClusterRef().index(), mkfitHits[ilay].size(), ilay, &hit);
+      mkfitHits[ilay].emplace_back(pos, err, totalHits);
+      ++totalHits;
+    }
+  }
+}
+
+mkfit::TrackVec MkFitInputConverter::convertSeeds(const edm::View<TrajectorySeed>& seeds,
+                                                  const MkFitIndexLayer& indexLayers,
+                                                  const TransientTrackingRecHitBuilder& ttrhBuilder,
+                                                  const MagneticField& mf) const {
+  mkfit::TrackVec ret;
+  ret.reserve(seeds.size());
+  int index = 0;
+  for (const auto& seed : seeds) {
+    const auto hitRange = seed.recHits();
+    const auto lastRecHit = ttrhBuilder.build(&*(hitRange.second - 1));
+    const auto tsos = trajectoryStateTransform::transientState(seed.startingState(), lastRecHit->surface(), &mf);
+    const auto& stateGlobal = tsos.globalParameters();
+    const auto& gpos = stateGlobal.position();
+    const auto& gmom = stateGlobal.momentum();
+    SVector3 pos(gpos.x(), gpos.y(), gpos.z());
+    SVector3 mom(gmom.x(), gmom.y(), gmom.z());
+
+    const auto cartError = tsos.cartesianError();  // returns a temporary, so can't chain with the following line
+    const auto& cov = cartError.matrix();
+    SMatrixSym66 err;
+    for (int i = 0; i < 6; ++i) {
+      for (int j = i; j < 6; ++j) {
+        err.At(i, j) = cov[i][j];
+      }
+    }
+
+    mkfit::TrackState state(tsos.charge(), pos, mom, err);
+    state.convertFromCartesianToCCS();
+    ret.emplace_back(state, 0, index, 0, nullptr);
+
+    // Add hits
+    for (auto iHit = hitRange.first; iHit != hitRange.second; ++iHit) {
+      const auto* hit = dynamic_cast<const BaseTrackerRecHit*>(&*iHit);
```

Comment:
Can't be avoided, `TrackingRecHit` does not provide information on the referred-to cluster, nor any information on the more concrete type so that `static_cast` could be used. On the other hand, the hit conversion consumes more time than seed conversion.

Reply:
Is this enough of a drag on performance to make it worth introducing an alternate TrajectorySeed class that only stores BaseTrackerRecHits?

Reply:
just reinterpret_cast and, if you want just to be sure that it is not junk, use one of trackerHitRTTI function

Reply:
Long-term plan is to avoid these data format conversions altogether, essentially by using the pixel hits and seeds from Patatrack in a commonly-agreed SoA(-like) format. I think additional effort for parallel data format optimization is not worth it.

(I also just checked some old profiles, the hit conversion takes 50x the time that the seed conversion takes)

Reply:
@VinInn 
> if you want just to be sure that it is not junk, use one of trackerHitRTTI function

Is `trackerHitRTTI::isFromDet()` the function which corresponds to `BaseTrackerRecHit`?

Reply:
in this specific case I would say yes...